Desnowflakes tank integrity

TemporalOroboros opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
- Makes tanks use obj_integrity instead of their own snowflaked version.
- Makes tanks check for exploding when they are destroyed, rather than once every process.
- Makes tanks always leak their gases when they are deconstructed.
- Removes the ability for tanks to seal themselves back up over time.
- Makes the bomb spawner actually produce functional bombs. 
- Removes the extraneous syndicate bomb spawner subtype.
- Miscellaneous code improvements to tanks, bomb spawners, and the blastcannon.
- Fixes the explosives compressor doubling the power of any bomb you put in it.

The changes to tank rupturing behavior shouldn't effect most tritium fueled TTVs including the 50K recipe. Toxins players don't need to worry about suddenly being incapable of getting points or refining anomaly cores. They should only really effect singlecaps, but I don't know enough about singlecaps to know what recipes I should test. I have confirmation that at least one mix is not effected by this.

The self-sealing properties of tanks have been removed. I'm not sure what the purpose of it was, I have heard that it was used to enable hand-portable plasmaflooding, but I'm not familiar with the practice.

As it turns out, the basic bomb spawners were broken on master! I have made the bombs they produce maxcap instead of just spring a leak. Since they maxcap by default now I have removed the syndicate subtype used to spawn the TTV produced by the cuban pete arcade game and replaced it with the normal timer subtype. Since none of the bomb spawner subtypes were used for anything else this shouldn't have any effect on the game.

On a similar note, I have discovered that the maxcap recipe on the wiki stopped working at some point since it was written. I will replace it with a functioning set of instructions.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
- Less snowflake code.
- Bomb spawners are actually functional now.
- Slightly better code.
- The explosives compressor accurately reflects the power of the bomb you put into it.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
code: Gas tanks now use object integrity rather than their own snowflaked version.
fix: Bomb spawners now produce functional bombs.
tweak: Tanks are no longer self-sealing.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

TemporalOroboros committed:
Makes tanks use object integrity

- Also makes blobs deconstruct items instead of directly qdeleting them

Jared-Fogle commented:
Please merge from upstream, there was a recent CI failure.

TemporalOroboros commented:
Alright.

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

Jared-Fogle commented on code/game/objects/items/tanks/tanks.dm:
```diff
@@ -213,9 +205,8 @@
-
-	check_status()
-	return 1
+	handle_tolerances(1)
```
What's with the fixed delta time?

TemporalOroboros replied:
It's because this is event based and I need some sort of scaling factor so extreme pressure/temperatures can actually have an effect.

Jared-Fogle replied:
Can you make this a define then instead of a magic number?

TemporalOroboros replied:
Sure

TemporalOroboros replied:
Done

Jared-Fogle commented on code/game/objects/items/tanks/tanks.dm:
```diff
@@ -228,59 +219,71 @@
+
+/// Handles the tank springing a leak.
+/obj/item/tank/obj_break(damage_flag)
+	if(!leaking)
+		leaking = TRUE
```
```suggestion
	leaking = TRUE
```

If it's already TRUE, then it just won't have any effect anyway.

TemporalOroboros replied:
Done.

TemporalOroboros committed:
Commits @Rohesie's suggestions. Also adds leaking and some minimal docs

- Makes leaking a simple `TRUE` assignment
- Adjusts the integrity failure point so that the tank actually starts leaking.
- Moves leaking to after the tolerance checks in `process(...)`

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

Jared-Fogle commented:
@LemonInTheDark Is this your domain?

Jared-Fogle commented:
Oh wait, it's things like water tanks, oops.

LemonInTheDark commented:
Very slightly, I would like to know how this change effects the amount of ticks a bomb is allowed to process for before blowing up, and how things scale, since it does appear to have changed that behavior, and it's very relevant to the strength of some explosions

LemonInTheDark commented:
I've talked a bit with oro, this needs a bit more time in the oven, some cleanup of stuff that make/rely on bombs, and some consideration about its balance effects. Has potential though, has a lot of potential

TemporalOroboros committed:
Minor code cleanup

- Uses [/obj/item/transfer_valve/proc/ready] for the blastcannon
- snake_case's some bombspawner variables.
- Tweaks the tanks bomb assembly code to be a tiny bit more sane.

TemporalOroboros commented:
Apparently several things related to tank explosions on master and the wiki are either outdated, broken, or both. I think I'll try to fix those as well.

TemporalOroboros committed:
Some more code cleanup. Fixes bomb spawners.

- Changes one instance of . = ..() to return ..()
- Rebalances the gases produced by bomb spawners.
- Removes the extraneous syndicate bomb spawner.

TemporalOroboros committed:
Slightly refactors the explosive compressor

- Makes the explosives compressor actually eject the raw core if it fails to compress it.

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

Jared-Fogle commented:
Can you merge from latest CI? I believe the test failure has been fixed.

TemporalOroboros commented:
I would if GitKraken ever acknowledged that the upstream has changed.

Jared-Fogle commented:
Make sure its fetched.

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

Jared-Fogle commented:
@LemonInTheDark Have they made the proper changes?

LemonInTheDark commented:
I think so, need to look a bit closer. On another note tho.
@TemporalOroboros Could you give me some numbers on different gasmixes and how many times they're allowed to react?

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

TemporalOroboros committed:
Merge remote-tracking branch 'upstream/master'

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

TemporalOroboros committed:
Leak warning flavor text

TemporalOroboros commented:
@LemonInTheDark 
Wiki 50K Mix for toxins still works.
1001kPa Oxygen + 99kPa Plasma at 385&deg;K cycled for about 10 seconds and then exploded for 1/2/4/0. Down from x/x/7/0 apparently.
500kPa Plasma + 500kPa Trit + 1000kPa Oxygen at 374&deg;K cycled for about 10 seconds and then non-explosively ruptured. I was told this used to be a maxcap. Adding a nitryl fuse just made it take longer to rupture.
208kPa Tritium + 192kPa Plasma + 400-600kPa Oxygen at &deg;K cycled for about 10 seconds and then sprung a leak and stopped gaining pressure. Supposedly this also used to be a maxcap.

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

TemporalOroboros committed:
Fixes location

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

TemporalOroboros committed:
Fixes whitespace

vincentiusvin commented:
> @LemonInTheDark
> Wiki 50K Mix for toxins still works.
> 1001kPa Oxygen + 99kPa Plasma at 385°K cycled for about 10 seconds and then exploded for 1/2/4/0. Down from x/x/7/0 apparently.
> 500kPa Plasma + 500kPa Trit + 1000kPa Oxygen at 374°K cycled for about 10 seconds and then non-explosively ruptured. I was told this used to be a maxcap. Adding a nitryl fuse just made it take longer to rupture.
> 208kPa Tritium + 192kPa Plasma + 400-600kPa Oxygen at °K cycled for about 10 seconds and then sprung a leak and stopped gaining pressure. Supposedly this also used to be a maxcap.

I should've been more clear on discord, but the third mix was supposed to be 530kpa Plasma, 220kpa trit, and 350 kpa oxygen. It maxcapped when i tried this build. (48/20/32, plasma/trit/oxy)
A maxcap nitryl recipe that i commonly use in master of 10kpa Oxy, 80kpa Trit, 710kpa Plasma, and 800kpa Nitryl (all room temp) also works fine and maxcaps.
I believe that there should be no problem with singlecaps caused by this PR if the recipe wasn't a hack to begin with.

However, one thing that might be worthy of note is the damage that is taken by the tank before the energetic trit reaction can occur. Usually when I accidentally filled too much oxygen in a singlecap tank it will explode in my face. With this build I instead get a leak. The delta_time variable might hike the integrity damage too much.

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

TemporalOroboros committed:
Merge master fixes

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

Whoneedspacee commented:
Seems fine to me, I don't see this changing the behavior of singlecaps either since their entire objective is to be low pressure and then spike, it might affect some recipes that toe the line energy wise but not really a big deal.

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

LemonInTheDark commented on code/game/machinery/computer/arcade.dm:
```diff
@@ -543,7 +543,7 @@ GLOBAL_LIST_INIT(arcade_prize_pool, list(
 			playsound(loc, 'sound/arcade/win.ogg', 50, TRUE)
 
 			if(obj_flags & EMAGGED)
-				new /obj/effect/spawner/newbomb/timer/syndicate(loc)
```
Is this meant to be weaker or something? I'm kinda confused about this typepath to be honest

TemporalOroboros replied:
The syndicate subtype was basically just a maxcap subtype for bomb spawners. (eg: It produced a maxcap rather than the weaker bombs the others did.) Since it was the only subtype that was actually used I nixed it and just made bomb spawners default to making maxcaps.

LemonInTheDark replied:
👍

LemonInTheDark commented on code/modules/research/anomaly/explosive_compressor.dm:
```diff
@@ -128,18 +124,21 @@
+	QDEL_NULL(inserted_bomb) // bomb goes poof
 	if(range < required_radius)
-		inserted_bomb.forceMove(src)
 		say("Resultant detonation failed to produce enough implosive power to compress [inserted_core]. Core ejected.")
-		return
-	QDEL_NULL(inserted_bomb) // bomb goes poof
+		inserted_core.forceMove(drop_location())
+		inserted_core = null
+		return FALSE
```
Behavior change?

TemporalOroboros replied:
It claimed to return the uncompressed core but never actually did so. Ifix.

LemonInTheDark replied:
👍

LemonInTheDark commented on code/game/objects/items/tanks/tanks.dm:
```diff
@@ -1,3 +1,12 @@
+/// How much time is assumed to pass while assuming air. Used to scale overpressure/overtemp damage when assuming air.
+#define ASSUME_AIR_DT_FACTOR 1
```
Is this in seconds?

TemporalOroboros replied:
Yes.

LemonInTheDark replied:
Could you make sure that's clear?

TemporalOroboros replied:
Done

LemonInTheDark commented on code/game/objects/items/tanks/tanks.dm:
```diff
@@ -235,59 +242,75 @@
+
+/// Handles the tank springing a leak.
+/obj/item/tank/obj_break(damage_flag)
+	if(!leaking)
+		leaking = TRUE
+		if(obj_integrity > 0) // So we don't play the alerts while we are exploding or rupturing.
+			visible_message("<span class='warning'>[src] springs a leak!</span>")
+			playsound(src, 'sound/effects/spray.ogg', 10, TRUE, -3)
```
Wouldn't it be cleaner to early return here?

TemporalOroboros replied:
Yes.

TemporalOroboros replied:
Done

TemporalOroboros commented on code/modules/research/anomaly/explosive_compressor.dm:
```diff
@@ -114,12 +114,8 @@
-	var/datum/gas_mixture/mix = new(70) // Standard tank volume, 70L
-	mix.merge(mix1)
-	mix.merge(mix2)
+	var/datum/gas_mixture/mix = new(0)
```
The TTV merges the tank volumes as well as the gas. The old way doubled the explosive power of normal TTVs and could reduce more efficient bombs by up to over a factor of 10.

LemonInTheDark replied:
👍

TemporalOroboros committed:
Merge branch 'master' into tank-integrity

TemporalOroboros committed:
Lemon's Changes + Arcade Fix

- Re-repaths the bomb spawner used by the cuban pete arcade machine emag
- Adds units to the doc for air assuming delta time
- Adds an early return to tank obj_break.

LemonInTheDark commented:
Gonna give this a quick run on live to make sure nothing major breaks, then we're good
